.github/workflows/e2e-tests.yaml

```diff
@@ -1,23 +1,17 @@
 name: End-to-End Testing
 
 on:
-  push:
-    # NOTE(antoineco): To avoid exposing repository secrets to user supplied
-    # code in pull requests, we only allow this workflow to run on the 'main'
-    # branch, where all code changes are expected to have been reviewed by
-    # maintainers.
-    #
-    # We may want to relax this in the future, on condition that safeguards are
-    # added to prevent the workflow from running automatically when an external
-    # contributor submits code changes inside the .github/workflows/ or
-    # test/e2e/ directories.
-    #
-    # Ref. https://securitylab.github.com/research/github-actions-preventing-pwn-requests/
-    branches: [ main ]
+  pull_request_review:
+    types: [submitted]
+    # Run the e2e test once the PR has been approved to allow for a final check before
+    # merging the PR. One of the contributers metioned in CODEOWNERS must approve the PR
+    # if it touches .github/ or test/e2e/ to prevent potential credential leakage.
```

Comment:
Note: this might make it difficult to get quick feedback when a PR is sent.

Let's take the example of #608, and let's assume that it was sent by someone who isn't a code owner.
The PR is broken to start with (in the context of the repo), but until someone leaves an approval (which might take days) you won't know that by looking at the CI status.

Generally speaking, if I send a PR adding a new test, and expect the CI to tell me whether it passes, I'll have to wait for an approval. But maybe reviewers want to base their decision on the outcome of the CI run, which leads to a chicken and egg situation.

It might not be a bad thing! But worth thinking about it.
Let's try and see if that workflow works for us.

Reply:
I'd agree, but this way we can ensure there are eyes on the PR in case they slipped in something malicious. I'm thinking of the "here's a clean PR, oh wait let me commit something grab the credentials, and then a follow-up commit to undo it" scenario. Which can become more troublesome if they squash their commits after the experiment and hide the attempt (except to show an extra push was made)

Reply:
GitHub thought about that. Any commit to the PR schedules a workflow run that needs to be approved (it happened to me recently while contributing a change to `ko`).
Approving one e2e run doesn't automatically approve subsequent runs.